core/runtime/src/main/java/io/quarkus/runtime/metrics/MetricsFactory.java

```diff
@@ -0,0 +1,141 @@
+package io.quarkus.runtime.metrics;
+
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.function.ToDoubleFunction;
+
+/**
+ * Extensions can create or register metrics using this factory
+ * independent of the enabled metrics provider
+ */
+public interface MetricsFactory {
+
+    /** A well-known string for MicroProfile metrics provided by the SmallRye Metrics quarkus extension */
+    final String MP_METRICS = "smallrye-metrics";
+
+    /** A well-known string for MicroProfile metrics provided by the SmallRye Metrics quarkus extension */
+    final String MICROMETER = "micrometer";
+
+    /**
+     * @return true if this factory supports the named metrics system. Arbitrary
+     *         strings are allowed. Constants are present for a few.
+     * @see #MICROMETER
+     * @see #MP_METRICS
+     */
+    boolean metricsSystemSupported(String name);
+
+    /** Create a fluid builder for registering counters and gauges */
+    MetricBuilder builder(String name);
+
+    interface MetricBuilder {
```

Comment:
I find it a bit odd when something called `*Builder` doesn't have any `build()` method to be called last, but it seems it would be quite hard to add one here because we build distinct types of objects. Just thinking, would it help if we renamed the `counter`, `gauge`, etc. methods to `buildCounter`, `buildGauge`?

Reply:
Naming things is hard. ;)